packages/runtime/datastore/src/dataStoreRuntime.ts

```diff
@@ -650,68 +650,141 @@ export class FluidDataStoreRuntime
-							0x17c /* "Unexpected attach (local) channel OP" */,
-						);
-					} else {
-						assert(!this.contexts.has(id), 0x17d /* "Unexpected attach channel OP" */);
-
-						const summarizerNodeParams = {
-							type: CreateSummarizerNodeSource.FromAttach,
-							sequenceNumber: message.sequenceNumber,
-							snapshot: attachMessage.snapshot,
-						};
-
-						const remoteChannelContext = this.createRemoteChannelContext(
-							attachMessage,
-							summarizerNodeParams,
-						);
-						this.contexts.set(id, remoteChannelContext);
+		for (const { message } of messagesWithMetadata) {
+			assert(message.type === messagesType, "Messages should have the same type");
+			try {
+				// catches as data processing error whether or not they come from async pending queues
+				switch (message.type) {
+					case DataStoreMessageType.ChannelOp:
+						// This is handled above. Break from the switch statement to emit "op" event.
+						break;
+					case DataStoreMessageType.Attach: {
+						const attachMessage = message.contents as IAttachMessage;
+						const id = attachMessage.id;
+
+						// We need to process the GC Data for both local and remote attach messages
+						processAttachMessageGCData(attachMessage.snapshot, (nodeId, toPath) => {
+							// Note: nodeId will be "/" unless and until we support sub-DDS GC Nodes
+							const fromPath = `/${this.id}/${id}${nodeId === "/" ? "" : nodeId}`;
+							this.dataStoreContext.addedGCOutboundRoute(fromPath, toPath, message.timestamp);
+						});
+
+						// If a non-local operation then go and create the object
+						// Otherwise mark it as officially attached.
+						if (local) {
+							assert(
+								this.pendingAttach.delete(id),
+								0x17c /* "Unexpected attach (local) channel OP" */,
+							);
+						} else {
+							assert(!this.contexts.has(id), 0x17d /* "Unexpected attach channel OP" */);
+
+							const summarizerNodeParams = {
+								type: CreateSummarizerNodeSource.FromAttach,
+								sequenceNumber: message.sequenceNumber,
+								snapshot: attachMessage.snapshot,
+							};
+
+							const remoteChannelContext = this.createRemoteChannelContext(
+								attachMessage,
+								summarizerNodeParams,
+							);
+							this.contexts.set(id, remoteChannelContext);
+						}
+						break;
 					}
-					break;
+					default:
 				}
 
-				case DataStoreMessageType.ChannelOp:
-					this.processChannelOp(message, local, localOpMetadata);
-					break;
-				default:
+				this.emit("op", message);
+			} catch (error) {
+				throw DataProcessingError.wrapIfUnrecognized(
+					error,
+					"fluidDataStoreRuntimeFailedToProcessMessage",
+					message,
+				);
 			}
+		}
+	}
 
-			this.emit("op", message);
-		} catch (error) {
-			throw DataProcessingError.wrapIfUnrecognized(
-				error,
-				"fluidDataStoreRuntimeFailedToProcessMessage",
+	/**
+	 * Process data store messages. The messages here are contiguous messages for this data store in a batch.
```

Comment:
`processBunchedMessages` and `processMessages` seems to have similar descriptions. Could we add some text that highlights how these two are different? Like processMessages is invoked by the layer above etc.

Reply:
`processBunchedMessages` is now removed.